common/error_test.cpp

```diff
@@ -29,94 +31,199 @@ auto IndirectError() -> Error { return Error("test"); }
 
 TEST(ErrorTest, IndirectError) { EXPECT_EQ(IndirectError().message(), "test"); }
 
-TEST(ErrorTest, ErrorOr) {
-  ErrorOr<int> err(Error("test"));
+TEST(ErrorTest, ErrorBuilderOperatorImplicitCast) {
+  ErrorOr<int> result = ErrorBuilder() << "msg";
+  EXPECT_THAT(result, IsError("msg"));
+}
 
-  EXPECT_THAT(err, IsError("test"));
+TEST(ErrorTest, StreamError) {
+  Error result = ErrorBuilder("TestFunc") << "msg";
+  RawStringOstream result_stream;
+  result_stream << result;
+  EXPECT_EQ(result_stream.TakeStr(), "TestFunc: msg");
 }
 
-TEST(ErrorTest, ErrorOrValue) { EXPECT_TRUE(ErrorOr<int>(0).ok()); }
+class CustomError : public ErrorBase<CustomError> {
+ public:
+  auto Print(llvm::raw_ostream& os) const -> void {
+    os << "Custom test error!";
+  }
+};
 
-auto IndirectErrorOrTest() -> ErrorOr<int> { return Error("test"); }
+template <typename ErrorT>
+struct ErrorOrTest : ::testing::Test {
```

Comment:
```suggestion
class ErrorOrTest : public ::testing::Test {
 public:
```

I thought we usually wrote tests as classes, because they have complex functionality?

I'll note too, for inheritance, we'd had a discussion about implicit public inheritance due to something zygoloid did and I think both Josh and I lacked the right instincts for `struct` -- we're too used to doing that on classes only.

Reply:
The test I copied this from that uses type parameterized testing used `struct`. 🤷🏻 Happy to switch here.